miden-lib/asm/miden/contracts/wallets/basic.masm

```diff
@@ -73,3 +73,165 @@ export.send_asset.1
+
+    # check whether ASSETS_HASH is an EMPTY_WORD ([0, 0, 0, 0]). In that case there are no assets 
+    # need to be added to the note
+    movdn.9 eqw
+    # => [flag, ASSETS_HASH, EMPTY_WORD, 0, note_idx, ...]
+
+    # if there are no assets to add
+    if.true
+        drop drop drop movup.6
+        # => [note_idx, 0, 0, EMPTY_WORD, ...]
+    else 
+        # get the note index to the top of the stack
+        movup.9
+
+        # get the number of assets 
+        # we need to divide the num_assets_felt by 4 to get the number of assets 
+        adv_push.1 div.4
+        # => [num_assets, note_idx, ASSETS_HASH, 0, EMPTY_WORD, ...]
+
+        # check if there is an odd number of assets
+        dup is_odd
+        # => [is_odd, num_assets, note_idx, ASSETS_HASH, 0, EMPTY_WORD, ...]
+
+        # copy is_odd, it defines if last asset word requires padding
+        dup movdn.3
+        # => [is_odd, num_assets, note_idx, needs_padding, ASSETS_HASH, 0, EMPTY_WORD, ...]
+
+        # get the even number of assets
+        sub
+        # => [num_assets_even, note_idx, needs_padding, ASSETS_HASH, 0, EMPTY_WORD, ...]
+
+        # `num_assets_even` will be used as a loop vaiable, so negate it to use add instead of sub 
+        # because the former uses one fewer cycle.
+        neg
+        # [-num_assets_even, note_idx, needs_padding, ASSETS_HASH, 0, EMPTY_WORD, ...]
+
+        # Prepare the capacity word. For rescue prime optimized the first element is
+        # set to `1` when padding is used and `0` otherwse, this is determined by the
+        # `needs_padding` flag.
+        dup.2 push.0.0.0
+        # => [CAPACITY, -num_assets_even, note_idx, needs_padding, ASSETS_HASH, 0, EMPTY_WORD, ...]
+
+        # set initial hasher state
+        padw padw
+        # => [B, A, CAPACITY, -num_assets_even, note_idx, needs_padding, ASSETS_HASH, 0, EMPTY_WORD, ...]
+
+        ### Iterate over the assets #######################################################
+        #
+        # Add assets to the note, simultaneously computing their accumulative hash 
+
+        # check loop condition
+        dup.12 neq.0
+        # => [b, B, A, CAPACITY, -num_assets_even, note_idx, needs_padding, ASSETS_HASH, 0, EMPTY_WORD, ...]
+
+        # while num_assets_even != 0
+        while.true
+            # load the assets from advice stack and remove them from the account 
+            adv_loadw exec.account::remove_asset
+            swapw adv_loadw exec.account::remove_asset
+            # => [ASSET_B, ASSET_A, CAPACITY, -num_assets_even, note_idx, needs_padding, ASSETS_HASH, 0, EMPTY_WORD, ...]
+
+            # add current asset to the note
+            dup.13 movdn.8
+            dupw dup.12
+            exec.tx::add_asset_to_note drop
+            swapw dupw movup.12
+            exec.tx::add_asset_to_note drop
+            swapw
+            # => [ASSET_B, ASSET_A, CAPACITY, -num_assets_even, note_idx, needs_padding, ASSETS_HASH, 0, EMPTY_WORD, ...]
```

Comment:
I would probably create a helper procedure - something like:
```
#! Inputs:  [note_id, ASSET, ...]
#! Outputs: [ASSET, ...]
proc.move_asset_to_note
```
And use it here and later on as well. It will take a few more cycles, but i think that's OK.

Also, why do we do `movup.12` instead of `dup.12` when we move the second asset?

Reply:
I'm using `movup` instead of `dup` because at that moment we were using the already duplicated note index on the position 8 (or 12, after the asset was duplicated), which should be dropped anyway. I can't always just duplicate the original `note_idx` since after asset duplication it moves out of the operating 16 element range.

```text
180: [ASSET_B, ASSET_A, note_idx, CAPACITY, -num_assets_even, note_idx, ...]
// in the lines 181 and 182 I could just move the duplicated `note_idx` back and forth, but I 
// thought that it will be cleaner and a little bit faster to just clone and drop it
181: [note_idx, ASSET_B, ASSET_B, ASSET_A, note_idx, CAPACITY, -num_assets_even, note_idx, ...]
182: [ASSET_B, ASSET_A, note_idx, CAPACITY, -num_assets_even, note_idx, ...]
//                         |                                     └─ original one that should stay 
//                         |                                        unchanged 
//                         └─ the duplicated one that we should drop anyway, so we can just 
//                            move it to the top of the stack 
183: [note_idx, ASSET_A, ASSET_A, ASSET_B, CAPACITY, -num_assets_even, note_idx, ...]
184: [ASSET_A, ASSET_B, CAPACITY, -num_assets_even, note_idx, ...]
185: [ASSET_B, ASSET_A, CAPACITY, -num_assets_even, note_idx, ...]
```

Actually, when I was writing this I stumbled over a thought "why can't we just use the actual `note_idx`, why should we duplicate it initially?". Indeed, we don't need to duplicate it, since it will be never dropped: `add_asset_to_note` retains the note index on the stack. I'll optimize it during the implementation of `send_note` and `move_asset_into_note`.